### PR TITLE
Improve ZedGraph label spacing

### PR DESCRIPTION
**Pull request by NateMG4**
Added functions to the GraphPane class that dynamically moves labels to avoid overlapping with other labels or points. The class FoldChangeVolcanoPlot uses these methods to make labels more readable on its scatter plot.

Labels Overlapping
image

Avoid Label Overlap
image

### CODE REVIEW

File: pwiz_tools/Skyline/Model/GroupComparison/GroupComparisonStrings.resx [hunk 1]
```diff
@@ -359,13 +359,16 @@
...
+    <comment>Example: Median Normalized Peak Area</comment>
   </data>
   <data name="EditGroupComparisonDlg_GetNormalizeOptionText_Batch_Calibration_Curve" xml:space="preserve">
     <value>Batch Calibration Curve</value>
   </data>
   <data name="EditGroupComparisonDlg_GroupComparisonDef_Group_comparisons_at_the_protein_level_using_calibrated_values_is_not_yet_supported_in_Skyline_" xml:space="preserve">
     <value>Group comparisons at the protein level using calibrated values is not yet supported in Skyline.</value>
   </data>
+  <data name="FoldChangeVolcanoPlot_BuildContextMenu_Avoid_Label_Overlap" xml:space="preserve">
+    <value>Avoid Label Overlap</value>
```
Comment (nickshulman) [inline pwiz_tools/Skyline/Model/GroupComparison/GroupComparisonStrings.resx:372]: I think the text of this menu item should be the same as the text on the checkbox in the Formatting dialog.
Currently, that checkbox says "Arrange labels" but I think everything should be changed to something like "Auto arrange labels".

File: pwiz_tools/Skyline/Controls/GroupComparison/FoldChangeVolcanoPlot.cs [hunk 2]
```diff
@@ -593,14 +600,23 @@ protected override void BuildContextMenu(ZedGraphControl sender, ContextMenuStri
...
+                }
+
+
             }
         }
 
+        private void OnLabelOverlapClick(object o, EventArgs eventArgs)
+        {
+            Settings.Default.GroupComparisonAvoidLabelOverlap = !Settings.Default.GroupComparisonAvoidLabelOverlap;
+            UpdateGraph();
```
Comment (nickshulman) [inline pwiz_tools/Skyline/Controls/GroupComparison/FoldChangeVolcanoPlot.cs:618]: It is possible for the user to have more than one volcano plot showing (if they have more than one Group Comparison) so, technically, you have to update all showing volcano plots when this setting changes.
It might be that the easiest way for all of these volcano plots to update themselves would be to subscribe to the "PropertyChanged" event on "Settings.Default".
Currently, there is no example code in Skyline for listening to property changes on Settings.Default, but I have been thinking for a long time that we should take advantage of that functionality. I usually add listeners like that in the OnHandleCreated event and remove them in the OnHandleDestroyed event.

**Review by nickshulman — COMMENTED** [review]
Looks very nice. Here are some comments.

File: pwiz_tools/Shared/zedgraph/ZedGraph/FontSpec.cs [hunk 3]
```diff
@@ -564,7 +564,7 @@ public FontSpec()
 		/// The Copy Constructor
 		/// </summary>
 		/// <param name="rhs">The FontSpec object from which to copy</param>
-		public FontSpec( FontSpec rhs )
+		public  FontSpec( FontSpec rhs )
```
Comment (brendanx67) [inline pwiz_tools/Shared/zedgraph/ZedGraph/FontSpec.cs:567]: Changes in this file seem unnecessary.

File: pwiz_tools/Shared/zedgraph/ZedGraph/GraphPane.cs [hunk 4]
```diff
@@ -1505,11 +1525,543 @@ public PieItem[] AddPieSlices( double[] values, string[] labels )
...
+            {
+                get => _label;
+                set
+                {
+                    _label = value;
+                    LabelPosition = _label.Location.TopLeft;
+                }
+            }
+
+            public LineObj Connector { get => _connector;
```
Comment (brendanx67) [inline pwiz_tools/Shared/zedgraph/ZedGraph/GraphPane.cs:1591]: Strange formatting

File: pwiz_tools/Shared/zedgraph/ZedGraph/GraphPane.cs [hunk 5]
```diff
@@ -1505,11 +1525,543 @@ public PieItem[] AddPieSlices( double[] values, string[] labels )
...
+                set
+                {
+					_connector = value;
+					ConnectorLoc = new Location(_connector.Location);
+                }
+            }
+			public VectorF LabelVector { get; set; }
```
Comment (brendanx67) [inline pwiz_tools/Shared/zedgraph/ZedGraph/GraphPane.cs:1598]: Strange formatting

File: pwiz_tools/Shared/zedgraph/ZedGraph/GraphPane.cs [hunk 6]
```diff
@@ -1505,11 +1525,543 @@ public PieItem[] AddPieSlices( double[] values, string[] labels )
...
+            public CurveItem Curve { get; set; }
+
+            public Location ConnectorLoc { get; private set; }
+			public PointF LabelPosition { get; private set; }
```
Comment (brendanx67) [inline pwiz_tools/Shared/zedgraph/ZedGraph/GraphPane.cs:1602]: Strange formatting

File: pwiz_tools/Shared/zedgraph/ZedGraph/GraphPane.cs [hunk 7]
```diff
@@ -1505,11 +1525,543 @@ public PieItem[] AddPieSlices( double[] values, string[] labels )
...
+            public bool IsSelected { get; private set; }
+
+			// This method is used to memorize starting positions during drags
```
Comment (brendanx67) [inline pwiz_tools/Shared/zedgraph/ZedGraph/GraphPane.cs:1605]: Make this a /// &lt;summary> comment

File: pwiz_tools/Shared/zedgraph/ZedGraph/GraphPane.cs [hunk 8]
```diff
@@ -1505,11 +1525,543 @@ public PieItem[] AddPieSlices( double[] values, string[] labels )
...
+            public void UpdateLabelLocation(double x, double y, GraphPane graph)
+            {
+				Label.Location.X = x;
+				Label.Location.Y = y;
+                using (var g = Graphics.FromHwnd(IntPtr.Zero))
+                {
+                    var newConnectorEnd = LabelLayout.CalculateConnectorSize(this, g, graph);
+                    Connector.Location.Width = newConnectorEnd.Width;
+                    Connector.Location.Height = newConnectorEnd.Height;
+                }
```
Comment (brendanx67) [inline pwiz_tools/Shared/zedgraph/ZedGraph/GraphPane.cs:1621]: Lots more strange formatting below. Please clean-up.

File: pwiz_tools/Shared/zedgraph/ZedGraph/GraphPane.cs [hunk 9]
```diff
@@ -21,13 +21,20 @@
...
 using System.Collections;
+using System.Collections.Generic;
 using System.Drawing.Imaging;
 using System.Windows.Forms;
 using System.Diagnostics;
 using System.IO;
 using System.Runtime.Serialization;
 using System.Security.Permissions;
 using System.ComponentModel;
+using System.Diagnostics.Eventing.Reader;
```
Comment (brendanx67) [inline pwiz_tools/Shared/zedgraph/ZedGraph/GraphPane.cs:32]: This is a lot of code to add to an already large and central class. Try to separate out most of this code into another file or more.

File: pwiz_tools/Shared/zedgraph/ZedGraph/TextObj.cs [hunk 10]
```diff
@@ -440,6 +440,7 @@ override public bool PointInBox( PointF pt, PaneBase pane, Graphics g, float sca
 						pts[2].X, pts[2].Y, pts[3].X, pts[3].Y );
 		}
 
+
```
Comment (brendanx67) [inline pwiz_tools/Shared/zedgraph/ZedGraph/TextObj.cs:443]: Unnecessary change

File: pwiz_tools/Skyline/Controls/Graphs/SummaryRelativeAbundanceGraphPane.cs [hunk 11]
```diff
@@ -388,6 +433,16 @@ private static bool IsStandard(PeptideDocNode pepDocNode)
         {
             return pepDocNode.GlobalStandardType != null;
         }
+        public void OnSuspendLayout(object sender, EventArgs eventArgs)
```
Comment (brendanx67) [inline pwiz_tools/Skyline/Controls/Graphs/SummaryRelativeAbundanceGraphPane.cs:440]: The handler should be private. It can wrap a public no-arguments function that can be called from tests. Normally, I would use something like:

SuspendLayout(!Settings.Default.GroupComparisonSuspendLabelLayout)

With:

public void SuspendLayout(bool suspend)
{
    Settings.Default.GroupComparisonSuspendLabelLayout = suspend;
    if (!suspend)
...

File: pwiz_tools/Skyline/Controls/GroupComparison/FoldChangeVolcanoPlot.cs [hunk 12]
```diff
@@ -182,6 +188,12 @@ protected override void OnShown(EventArgs e)
             }
         }
 
+        protected override void OnHandleCreated(EventArgs e)
```
Comment (brendanx67) [inline pwiz_tools/Skyline/Controls/GroupComparison/FoldChangeVolcanoPlot.cs:191]: Couldn't this be done in OnShown() with the other event handlers?

File: pwiz_tools/Skyline/Controls/GroupComparison/FoldChangeVolcanoPlot.cs [hunk 13]
```diff
@@ -432,7 +447,7 @@ private void BindingListSourceOnListChanged(object sender, ListChangedEventArgs
             QueueUpdateGraph();
         }
 
-        private bool zedGraphControl_MouseMoveEvent(ZedGraphControl sender, MouseEventArgs e)
+        private bool  zedGraphControl_MouseMoveEvent(ZedGraphControl sender, MouseEventArgs e)
```
Comment (brendanx67) [inline pwiz_tools/Skyline/Controls/GroupComparison/FoldChangeVolcanoPlot.cs:450]: Unnecessary change

File: pwiz_tools/Skyline/Controls/GroupComparison/FoldChangeVolcanoPlot.cs [hunk 14]
```diff
@@ -589,23 +618,53 @@ protected override void BuildContextMenu(ZedGraphControl sender, ContextMenuStri
...
                 if (AnyCutoffSettingsValid)
                 {
                     menuStrip.Items.Insert(index++, new ToolStripSeparator());
                     menuStrip.Items.Insert(index, new ToolStripMenuItem(GroupComparisonStrings.FoldChangeVolcanoPlot_BuildContextMenu_Remove_Below_Cutoffs, null, OnRemoveBelowCutoffsClick));
-                }  
+                }
             }
         }
 
+        public void OnLabelOverlapPropertyChange(object sender, PropertyChangedEventArgs e)
```
Comment (brendanx67) [inline pwiz_tools/Skyline/Controls/GroupComparison/FoldChangeVolcanoPlot.cs:643]: Event handlers should not be made public. Instead define a simpler public function that the private event handler calls

File: pwiz_tools/Skyline/Controls/GroupComparison/FoldChangeVolcanoPlot.cs [hunk 15]
```diff
@@ -589,23 +618,53 @@ protected override void BuildContextMenu(ZedGraphControl sender, ContextMenuStri
...
+            Settings.Default.GroupComparisonAvoidLabelOverlap = !Settings.Default.GroupComparisonAvoidLabelOverlap;
+        }
         private void OnFormattingClick(object o, EventArgs eventArgs)
         {
             ShowFormattingDialog();
         }
 
+        private void OnSuspendLayout(object sender, EventArgs eventArgs)
+        {
+            Settings.Default.GroupComparisonSuspendLabelLayout = !Settings.Default.GroupComparisonSuspendLabelLayout;
```
Comment (brendanx67) [inline pwiz_tools/Skyline/Controls/GroupComparison/FoldChangeVolcanoPlot.cs:660]: See comment on this function in the abundance plot. That it is private implies that it is not tested here.

**Review by brendanx67 — CHANGES_REQUESTED** [review]
I would like to see a bit of code clean-up here. Hopefully it won't take very long.

**Comment by brendanx67** [comment]
More testing: GraphPane.IsOverLabel() only works if auto-arrange is on. If you turn it off, then you will no longer see a hand cursor or be able to click on labels to select a point in the relative abundance plot.

**Review by brendanx67 — APPROVED** [review]
